include/glow/ExecutionEngine/ExecutionEngine2.h

```diff
@@ -0,0 +1,163 @@
+ */
+#ifndef GLOW_EXECUTIONENGINE_EXECUTIONENGINE2_H
+#define GLOW_EXECUTIONENGINE_EXECUTIONENGINE2_H
+
+#include "glow/Backend/Backend.h"
+#include "glow/Backend/CompiledFunction.h"
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Base/Train.h"
+#include "glow/Base/Traits.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/PlaceholderBindings.h"
+#include "glow/Runtime/HostManager/HostManager.h"
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include <memory>
+#include <unordered_map>
+
+namespace glow {
+
+/// This is the ExecutionEngine. It owns the Graph, the backend, and the
+/// compiled function.  The Graph, etc in this class are defined as pointers, in
+/// order to erase the type and prevent the internal types from leaking out to
+/// the users of this class.
+class ExecutionEngine final {
+  /// Module containing the function and supporting information.
+  std::unique_ptr<Module> module_;
+
+  /// Raw pointer to module_ this is to support module access after the module
+  /// has been added to hostManager_.
+  Module *rawModule_;
+
+  /// The network execution backend.
+  Backend *backend_ = nullptr;
+
+  /// Whether or not the ExecutionEngine owns the backend or is just using
+  /// a backend provided from elsewhere. If ownsBackend is true,
+  /// ~ExecutionEngine will delete the backend_.
+  bool ownsBackend_ = false;
+
+  /// The HostManager for executing the compiled functions.
+  std::unique_ptr<runtime::HostManager> hostManager_;
+
+  /// Glow functions compiled for this ExecutionEngine's backend.
+  std::set<std::string> compiledFunctions_;
+
+  /// Single execution of the given \compiledFunction with the given context
+  /// \bindings.
+  void runInternal(ExecutionContext &context, llvm::StringRef name);
+
+public:
+  ExecutionEngine(llvm::StringRef backend = "Interpreter");
+
+  ~ExecutionEngine();
+
+  /// Set the code generator to \p backend. New code will be generated
+  /// using this backend. This clears all previously loaded functions and resets
+  /// the Module.
+  void setBackend(llvm::StringRef backend);
+
+  /// Set the code generator to a custom \p backend. If \p ownsBackend is false
+  /// then ExecutionEngine will use the given backend without owning it which
+  /// means that ~ExecutionEngine will not delete it.
+  void setBackend(Backend *backend, bool ownsBackend = true);
+
+  /// Get a pointer to the backend.
+  const Backend *getBackend() const;
+
+  /// \returns the internal graph.
+  Module &getModule() { return *rawModule_; }
+
+  /// Clears the ExecutionEngine and all CompiledFunctions.
+  void clear();
+
+  /// \returns the DAG for the specified \p network.
+  llvm::Expected<runtime::DAG &> getDAG(llvm::StringRef network) {
+    return hostManager_->getNetworkDAG(network);
+  }
+
+  /// \returns whether a node with the provided \p NI is supported by the
+  /// underlying backend.
+  bool isOpSupported(const NodeInfo &NI) const {
+    return backend_->isOpSupported(NI);
```

Comment:
E.g. this is an example of a call that can be wrong if this function is not static since a different Backend object is used for compile.

Reply:
also gone :)